src/rpc/zmq_pub.cpp

```diff
@@ -0,0 +1,462 @@
+  {
+    return lhs.name < rhs;
+  }
+
+  struct is_valid
+  {
+    bool operator()(const cryptonote::txpool_event& event) const noexcept
+    {
+      return event.valid;
+    }
+  };
+
+  template<typename T, std::size_t N>
+  void verify_sorted(const std::array<context<T>, N>& elems, const char* name)
+  {
+    auto unsorted = std::is_sorted_until(elems.begin(), elems.end());
+    if (unsorted != elems.end())
+      throw std::logic_error{name + std::string{" array is not properly sorted, see: "} + unsorted->name};
+  }
+
+  void write_header(rapidjson::StringBuffer& buf, const boost::string_ref name)
+  {
+    PutReserve(buf, name.size());
+    for (const char ch : name)
+      PutUnsafe(buf, ch);
+    buf.Put(':');
+  }
+
+  //! \return `name:...` where `...` is JSON and `name` is directly copied (no quotes - not JSON).
+  template<typename T>
+  void json_pub(rapidjson::StringBuffer& buffer, const T value)
+  {
+    rapidjson::Writer<rapidjson::StringBuffer> dest{buffer};
+    using cryptonote::json::toJsonValue;
+    toJsonValue(dest, value);
+  }
+
+  //! Object for "minimal" block serialization
+  struct minimal_chain
+  {
+    const std::uint64_t height;
+    const cryptonote::block& block;
+  };
+
+  //! Object for "minimal" tx serialization
+  struct minimal_txpool
+  {
+    const cryptonote::transaction& tx;
+  };
+
+  void toJsonValue(rapidjson::Writer<rapidjson::StringBuffer>& dest, const minimal_chain self)
+  {
+    crypto::hash id{};
+    if (!get_block_hash(self.block, id))
+    {
+      MERROR("ZMQ/Pub failure: get_block_hash");
+      return;
+    }
+
+    dest.StartObject();
+    INSERT_INTO_JSON_OBJECT(dest, height, self.height);
+    INSERT_INTO_JSON_OBJECT(dest, prev_id, self.block.prev_id);
+    INSERT_INTO_JSON_OBJECT(dest, id, id);
+    dest.EndObject();
+  }
+
+  void toJsonValue(rapidjson::Writer<rapidjson::StringBuffer>& dest, const minimal_txpool self)
+  {
+    crypto::hash id{};
+    std::size_t blob_size = 0;
+    if (!get_transaction_hash(self.tx, id, blob_size))
+    {
+      MERROR("ZMQ/Pub failure: get_transaction_hash");
+      return;
+    }
+
+    dest.StartObject();
+    INSERT_INTO_JSON_OBJECT(dest, id, id);
+    INSERT_INTO_JSON_OBJECT(dest, blob_size, blob_size);
+    dest.EndObject();
+  }
+
+  void json_full_chain(rapidjson::StringBuffer& buf, const std::uint64_t height, const cryptonote::block& bl)
+  {
+    json_pub(buf, std::cref(bl));
+  }
+
+  void json_minimal_chain(rapidjson::StringBuffer& buf, const std::uint64_t height, const cryptonote::block& bl)
+  {
+    json_pub(buf, minimal_chain{height, bl});
+  }
+
+  // boost::adaptors are in place "views" - no copy/move takes place
+  // moving transactions (via sort, etc.), is expensive!
+
+  void json_full_txpool(rapidjson::StringBuffer& buf, epee::span<const cryptonote::txpool_event> txes)
+  {
+    namespace adapt = boost::adaptors;
+    const auto to_full_tx = [](const cryptonote::txpool_event& event)
+    {
+      return event.tx;
```

Comment:
It just occurred to me that the full txes don't have the hash included either. They must be computed remotely by duplicating the daemon code. Worth sending the hash here too?